discovery_engine_core/providers/src/lib.rs

```diff
@@ -31,23 +31,180 @@
+    /// Url path for the news search provider.
+    pub news_provider_path: String,
+    /// Url path for the latest headlines provider.
+    pub headlines_provider_path: String,
+}
+
+pub struct Providers {
+    pub headlines: Arc<dyn HeadlinesProvider>,
+    pub trusted_headlines: Arc<dyn TrustedHeadlinesProvider>,
+    pub news: Arc<dyn NewsProvider>,
+    pub trending_topics: Arc<dyn TrendingTopicsProvider>,
+}
+
+fn create_endpoint_url(str_base_url: &str, path: &str) -> Result<Url, Error> {
+    let mut base_url = Url::parse(str_base_url).map_err(|_| Error::MalformedUrlInConfig {
+        url: str_base_url.into(),
+    })?;
+
+    let mut segments = base_url
+        .path_segments_mut()
+        .map_err(|_| Error::MalformedUrlInConfig {
+            url: str_base_url.into(),
+        })?;
+
+    segments.pop_if_empty();
+    let stripped_path = path.strip_prefix('/').unwrap_or(path);
+    let stripped_path = stripped_path.strip_suffix('/').unwrap_or(stripped_path);
+
+    for new_segment in stripped_path.split('/') {
+        segments.push(new_segment);
+        if new_segment.is_empty() {
+            return Err(Error::MalformedUrlPathInConfig { path: path.into() });
+        }
+    }
+
+    drop(segments);
+    Ok(base_url)
+}
+
+fn select_provider_impl<T: ?Sized>(
+    endpoint: RestEndpoint,
+    create_newscatcher: impl FnOnce(RestEndpoint) -> Arc<T>,
+) -> Result<Arc<T>, Error> {
+    if let Some(segments) = endpoint.url().path_segments() {
+        for segment in segments {
+            return match segment {
+                "newscatcher" => Ok(create_newscatcher(endpoint)),
+                _ => continue,
+            };
+        }
+    }
+
+    Err(Error::NoProviderForEndpoint {
+        url: endpoint.url().to_string(),
+    })
+}
+
+impl Providers {
+    pub fn new(config: &ProviderConfig) -> Result<Self, Error> {
+        let headlines_endpoint = RestEndpoint::new(
+            create_endpoint_url(&config.api_base_url, &config.headlines_provider_path)?,
+            config.api_key.clone(),
+        );
+
+        let headlines = select_provider_impl(
+            headlines_endpoint,
+            NewscatcherHeadlinesProvider::from_endpoint,
+        )?;
+
+        let news_endpoint = RestEndpoint::new(
+            create_endpoint_url(&config.api_base_url, &config.news_provider_path)?,
+            config.api_key.clone(),
+        );
+
+        let news = select_provider_impl(news_endpoint, NewscatcherNewsProvider::from_endpoint)?;
+
+        //Note: Trusted-sources only works with newscatcher for now.
+        let trusted_headlines_endpoint = RestEndpoint::new(
+            create_endpoint_url(&config.api_base_url, "newscatcher/v2/trusted-sources")?,
+            config.api_key.clone(),
+        );
+        let trusted_headlines =
+            NewscatcherTrustedHeadlinesProvider::from_endpoint(trusted_headlines_endpoint);
+
+        let trending_topics_endpoint = RestEndpoint::new(
+            // Why is this just plain `_tt`?
```

Comment:
Because we don't have a `bing/v1/trending-topics` endpoint.

Reply:
I've changed the endpoint to what we want to use in the future, and opened TY-2995 so that we can keep track of that.